### PR TITLE
feat: improve our AJV usage

### PR DESCRIPTION
**Pull request by P0lip**
### Please check if the PR fulfills these requirements
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

Feature (fewer messages?) + fixes for https://github.com/stoplightio/spectral/issues/187 https://github.com/stoplightio/spectral/issues/116

### What is the current behavior?
image

### If this is a feature change, what is the new behavior?
image

### Does this PR introduce a breaking change?

Not really :)

### Other information

My initial intention was to make CLI messages more flexible, but this turned out to be not the greatest idea as we actually need meaningful messages only for stylish CLI format. API usage and JSON CLI format should be untouched, so decided to add the simple condition, so that both worlds are satisfied.

(e.g. detailed explanation, related issues, links for us to have context, eg. stackoverflow, issues outside of the repo, forum, etc.)

### CODE REVIEW

File: src/functions/schema.ts [hunk 1]
```diff
@@ -14,6 +15,12 @@ ajv._opts.defaultMeta = jsonSpecv4.id;
 // @ts-ignore
 ajv._refs['http://json-schema.org/schema'] = 'http://json-schema.org/draft-04/schema';
 
+ajv.addFormat('int32', { type: 'number', validate: validator.int32 });
+ajv.addFormat('int64', { type: 'number', validate: validator.int64 });
+ajv.addFormat('float', { type: 'number', validate: validator.float });
+ajv.addFormat('double', { type: 'number', validate: validator.double });
+ajv.addFormat('byte', { type: 'string', validate: validator.byte });
```
Comment (XVincentX) [inline src/functions/schema.ts]: I had a discussion with @philsturgeon about this just yesterday and he was against this. Did you folks change your mind?

**Comment by marbemac** [comment]
Another thought, might be worth evaluating https://github.com/atlassian/better-ajv-errors.

File: src/functions/schema.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import { decodePointerFragment } from '@stoplight/json';
 import * as AJV from 'ajv';
 import * as jsonSpecv4 from 'ajv/lib/refs/json-schema-draft-04.json';
+const validator = require('ajv-oai/lib/format-validator');
```
Comment (philsturgeon) [inline src/functions/schema.ts]:
```suggestion
const oasFormatValidator = require('ajv-oai/lib/format-validator');
```

File: src/functions/schema.ts (hunk 1 above)
Comment (philsturgeon) [reply src/functions/schema.ts]: Ok, this approach is a bit nicer than what I had in mind when we discussed using ajv-oas. I thought we'd be passing all validation through that function, instead we're kinda teaching our generic validation logic about some of the formats OpenAPI people are likely to use, and seeing as OpenAPI and JSON Schema are used interchangeably by many its not hurting to have these available in generic JSON Schema usage anyway. 

Let's go with this, and if we get pushback about having OpenAPI-specific formats enabled for all the things then... ok we can look into adding specific validation rules for specific rulesets, or autodetecting OpenAPI, or some other thing.

**Comment by P0lip** [comment]
> I would like to see our summary generally be more useful, it should be a very quick hint as to what is going on, with a little information about the field or whatever. Those in the know will look at this short summary and know how to fix it, but others will want more information.

I'd drop summary entirely. Not really sure what's the point of that property.
Ideally, we should have something similar to what ESLint does. That is:
```

I'm not sure whether it makes sense to include property given the fact path is already present.
There is path available

**Comment by P0lip** [comment]
> I would like to see our summary generally be more useful, it should be a very quick hint as to what is going on, with a little information about the field or whatever. Those in the know will look at this short summary and know how to fix it, but others will want more information.

I'd drop summary entirely. Not really sure what's the point of that property.
Ideally, we should have something similar to the snippet below (inspired by ESLint)
```
{
    rule: "oas2-schema",
    message: "Expected something to be truthy",
    path: ['some','path']
    severity: 2,
    line: 100,
    column: 2
}
```
I'm not sure whether it makes sense to include property given the fact path is already present.
It'll also make things more difficult in Studio as we already do highlight the line and point to given line and column etc, so we would most likely want to avoid any long messages.
I'm open to discussion though. I do agree that the current solution is not ideal.

> This more information might have strings we've written or other variables involved, with maybe information on how to fix it, etc.

Agreed, would be nice to give some feedback. I'd say the best way to do it is to provide an extensive doc about what given rule does and give examples when it is met and when not.
Yet again, ESLint is a good example here
https://eslint.org/docs/rules/
https://eslint.org/docs/rules/eqeqeq
https://eslint.org/docs/rules/no-unexpected-multiline

File: src/functions/schema.ts (hunk 1 above)
Comment (XVincentX) [reply src/functions/schema.ts]: Hmm, ok. I fail to understand the difference with what was proposed last week though. It was the exact same solution.

**Comment by philsturgeon** [comment]
Literally putting everything through the os validator, vrs adding a few extra rules from that validator, is conceptually different enough that it doesn’t feel quite as unsettling.  

--
Phil Sturgeon
@philsturgeon

**Comment by marbemac** [comment]
> I'd drop summary entirely.

I agree, summary doesn't make much sense on the validation results (still makes sense on the rule definition). Instead, consumers could retrieve summary on demand, as needed. We already expose `get rules()` on the spectral instance, so with no changes a consumer could do this:

```ts
const summary = spectralInstance.rules[result.rule].summary;
```

> I'm not sure whether it makes sense to include property given the fact path is already present.

Yeah I agree in general, but there is a tricky use case - how to describe a missing property. Because the property does not exist in the source, we have nothing to highlight / point to inside of Studio, for example. So we end up putting the error on the parent object, and having a bit more information about the missing property in the message is helpful, e.g. "the description property is missing" versus "a property is missing" (and having user look at the path to see which property is missing).

> I would like to see our summary generally be more useful, it should be a very quick hint as to what is going on, with a little information about the field or whatever. Those in the know will look at this short summary and know how to fix it, but others will want more information.

Maybe we should more formally define what the goal of `message` (property on each validation result) and `summary` (property on the rule itself) is? As it stands right now, message is generated at runtime, during validation, and summary is hardcoded into the rule itself and never changes.

> That said, I'm still open to making messages more customizable.
We could implement a decent string interpolation.

We could run messages through https://mustache.github.io, and expose several variables to it, so that rule authors can generate more useful messages?

> Literally putting everything through the os validator

I agree - we should improve the messages in spectral itself (which will benefit all downstream consumers, such as Studio), versus cleaning things up in the CLI (assuming I'm understanding this PR correctly).

**Comment by P0lip** [comment]
> Maybe we should more formally define what the goal of message (property on each validation result) and summary (property on the rule itself) is? As it stands right now, message is generated at runtime, during validation, and summary is hardcoded into the rule itself and never changes.

`summary` could be renamed to `description`. This would be less confusing.

> Yeah I agree in general, but there is a tricky use case - how to describe a missing property. Because the property does not exist in the source, we have nothing to highlight / point to inside of Studio, for example. So we end up putting the error on the parent object, and having a bit more information about the missing property in the message is helpful, e.g. "the description property is missing" versus "a property is missing" (and having user look at the path to see which property is missing).

This is indeed a valid case.

> We could run messages through https://mustache.github.io, and expose several variables to it, so that rule authors can generate more useful messages?

Frankly speaking, this what I wanted to initially implement in the scope of this PR.
I even implemented some code that should be sufficient to cover our needs.
```

**Comment by P0lip** [comment]
@chris-miaskowski you might want to take a look at this PR as well

**Comment by philsturgeon** [comment]
This PR is in response to https://stoplightio.atlassian.net/browse/SO-84 which is meant to formally define what summary and message are, because sometimes summary is useful and sometimes message is useful. Having the two has lead to inconsistencies.

I feel like we got to a good consensus here, lets drop summary entirely and just have message, which looks like this: 

image

Basically exactly what you've done, but everywhere, for all the formatters. 

To avoid needing to make this a breaking change and therefore v3.0, we can just make summary an alias of message and mark it as deprecated in code, or something like that?

**Comment by P0lip** [comment]
@philsturgeon 
Sounds like a good plan to me! Will do.

File: src/rulesets/oas3/__tests__/valid-example.ts [hunk 3]
```diff
@@ -3,6 +3,8 @@ import { oas3Rules } from '../index';
 
 const ruleset = { rules: oas3Rules() };
 
+declare var test: jest.It;
```
Comment (XVincentX) [inline src/rulesets/oas3/__tests__/valid-example.ts]: Why is this required?

File: src/__tests__/linter.test.ts [hunk 4]
```diff
@@ -206,19 +206,16 @@ responses:: !!foo
     expect(result).toEqual([
       expect.objectContaining({
         code: 'oas3-schema',
-        summary: 'Validate structure of OpenAPIv3 specification.',
```
Comment (XVincentX) [inline src/__tests__/linter.test.ts]: If `summary` is the same as `message`, wouldn't it be worth to still check its presence?

File: src/linter.ts [hunk 5]
```diff
@@ -88,13 +88,25 @@ export const lintNode = (
...
-        const location = parsedResult.getLocationForJsonPath(parsed, result.path || targetPath, true);
+        const path = result.path || targetPath;
+        const location = parsedResult.getLocationForJsonPath(parsed, path, true);
 
         return {
           code: rule.name,
-          summary: rule.summary,
-          message: result.message,
-          path: result.path || targetPath,
+          get summary() {
```
Comment (XVincentX) [inline src/linter.ts]: Nitpick: new line

File: src/linter.ts [hunk 6]
```diff
@@ -88,13 +88,25 @@ export const lintNode = (
...
+            return this.message;
+          },
+          message:
```
Comment (XVincentX) [inline src/linter.ts]: Nitpick: new line

File: src/linter.ts (hunk 5 above)
Comment (XVincentX) [reply src/linter.ts]: I guess this is a property because we can't use `this` if not in a defined function, right?

File: src/linter.ts (hunk 6 above)
Comment (XVincentX) [reply src/linter.ts]: Ok took me some time to get this — maybe a comment could help here

File: src/rulesets/oas3/__tests__/valid-example.ts (hunk 3 above)
Comment (P0lip) [reply src/rulesets/oas3/__tests__/valid-example.ts]: We have `mocha` test runner included in Spectral and TS is confused and doesn't know which `it` function it should refer to.

Comment (philsturgeon) [reply src/rulesets/oas3/__tests__/valid-example.ts]: Can we get rid of mocha? I think oclif added jest support a while back. https://github.com/oclif/oclif/pull/219

Comment (P0lip) [reply src/rulesets/oas3/__tests__/valid-example.ts]: Sorry, was wrong. Mocha is not listed as a dependency, but TS still sees typings for it and prefers it over Jest.

Comment (P0lip) [reply src/rulesets/oas3/__tests__/valid-example.ts]:
```
# npm ls @types/mocha # gives the following
@stoplight/spectral@0.0.0 /home/Stoplight/spectral
└─┬ @oclif/test@1.2.4
  └─┬ fancy-test@1.4.3
    └── @types/mocha@5.2.6 
```
Seems like oclif/test requires types/mocha

File: src/__tests__/linter.test.ts (hunk 4 above)
Comment (P0lip) [reply src/__tests__/linter.test.ts]: Sounds reasonable.

File: src/rulesets/oas3/__tests__/valid-example.ts (hunk 3 above)
Comment (P0lip) [reply src/rulesets/oas3/__tests__/valid-example.ts]: Well, they haven't released any newer version, so we'll need to stick to that unless you have a better workaround.
The reason I had to override the type is caused by the use of `test.each`. Plain `it` or `test` have a very similar, if not exact, signature.
Should I add a comment explaining that?
@XVincentX @philsturgeon

Comment (XVincentX) [reply src/rulesets/oas3/__tests__/valid-example.ts]: Ok figured it out in person. We can't do anything about this unless we decide to ditch the whole oclif testing framework.

File: src/rulesets/oas3/__tests__/valid-example.ts [hunk 7]
```diff
@@ -3,6 +3,8 @@ import { oas3Rules } from '../index';
 
 const ruleset = { rules: oas3Rules() };
 
+// @oclif/test packages requires @types/mocha, therefore we have 2 packages coming up with similar typings
+// TS is confused and prefers the mocha ones, so we need to instrument it to pick up the Jest ones
```
Comment (XVincentX) [inline src/rulesets/oas3/__tests__/valid-example.ts]: Nitpick: newline.

**Comment by P0lip** [comment]
@philsturgeon do you have any extra feedback or can I merge it?

@marbemac a few words on the package you mentioned - it seems truly great. I'm not sure what others think, but IMHO it might be a nice addition. Just wondering whether we should do it as a part of this PR or next one. Up to discussion.

**Comment by marbemac** [comment]
Totally cool to do a second pass and separate PR to make the messages even better in the future! Don't need to block this one.

File: src/rulesets/oas/functions/oasPathParam/__tests__/__snapshots__/index.test.ts.snap [hunk 8]
```diff
@@ -33,11 +31,7 @@ exports[`oasPathParam Error if duplicate path parameters with same name are used
 Array [
   Object {
     "code": "path-params",
-    "message": "The path \\"**/foo/{bar}/{bar}**\\" uses the parameter \\"**{bar}**\\" multiple times.
-
-Path parameters must be unique.
-
-To fix, update the path so that all parameter names are unique.",
```
Comment (philsturgeon) [inline src/rulesets/oas/functions/oasPathParam/__tests__/__snapshots__/index.test.ts.snap]: The old message said they should be unique and the new one doesn’t. Can we bring that back?

Comment (P0lip) [reply src/rulesets/oas/functions/oasPathParam/__tests__/__snapshots__/index.test.ts.snap]: Sure Phil, thanks for spotting this, I'll create a PR to address that soon (certainly today, this is 21.05).
